[eks/cluster] Update to use AWS Auth API

Nuru opened the pull request:
## Breaking Changes

>[!WARNING]
> This release contains breaking changes to the `eks/cluster` component.
Read the migration guidance in the [CHANGELOG](./modules/eks/cluster/CHANGELOG.md)

## what

- Upgrade the `eks/cluster` component to use the AWS API for access control

## why

- The old mechanism, using a ConfigMap, was unreliable

## references

- `terraform-aws-eks-cluster` [v4.0.0 Release Notes](https://github.com/cloudposse/terraform-aws-eks-cluster/releases/tag/4.0.0)
- `terraform-aws-eks-cluster` [v4 Migration Guide](https://github.com/cloudposse/terraform-aws-eks-cluster/blob/main/docs/migration-v3-v4.md)
- [New API for EKS access control](https://github.com/aws/containers-roadmap/issues/185#issuecomment-1863025784)

dudymas commented on modules/eks/cluster/CHANGELOG.md:
```diff
@@ -20,13 +220,14 @@ single node, removing the high availability that comes from having a node per Av
+**NOTE about instance type**: The `c7a.medium` instance type is relatively new. If you have deployed an old version of
+our [ServiceControlPolicy](https://github.com/cloudposse/terraform-aws-service-control-policies)
+`DenyEC2NonNitroInstances`, `DenyNonNitroInstances` (obsolete, replaced by `DenyEC2NonNitroInstances`), and/or
+`DenyEC2InstancesWithoutEncryptionInTransit`, you will want to update them to v0.14.1 or choose a difference instance
```
```suggestion
`DenyEC2InstancesWithoutEncryptionInTransit`, you will want to update them to v0.14.1 or choose a different instance
```

dudymas commented on modules/eks/cluster/main.tf:
```diff
@@ -10,48 +10,75 @@ locals {
-    [local.karpenter_role_arn]
+    # As of Karpenter v0.35.0, there is no entry in the official Karpenter documentation
+    # stating how to configure Karpenter node roles via EKS Access Entries.
+    # However, it is launching unmanaged worker nodes, so it makes sense they they
```
```suggestion
    # However, it is launching unmanaged worker nodes, so it makes sense that they
```

dudymas commented on modules/eks/cluster/main.tf:
```diff
@@ -10,48 +10,75 @@ locals {
+    # As of Karpenter v0.35.0, there is no entry in the official Karpenter documentation
+    # stating how to configure Karpenter node roles via EKS Access Entries.
+    # However, it is launching unmanaged worker nodes, so it makes sense they they
+    # be configured as EC2_LINUX unmanaged worker nodes. Of course, this does probably
```
```suggestion
    # be configured as EC2_LINUX unmanaged worker nodes. Of course, this probably
```

dudymas commented on modules/eks/cluster/variables.tf:
```diff
@@ -93,27 +93,11 @@ variable "cluster_log_retention_period" {
-  nullable    = false
-}
-
+# TODO:
```
have we made tickets for these?

Nuru replied:
No. We will do it when we need it. So far, none of our customers need it.

dudymas commented on modules/eks/cluster/main.tf:
```diff
@@ -10,48 +10,75 @@ locals {
+  access_entry_map = merge(local.aws_team_roles_access_entry_map, local.aws_sso_access_entry_map, local.iam_roles_access_entry_map, local.iam_users_access_entry_map)
 
-  # If Karpenter IAM role is enabled, add it to the `aws-auth` ConfigMap to allow the nodes launched by Karpenter to join the EKS cluster
+  # Existing Fargate Profile role ARNs
+  fargate_profile_role_arns = local.eks_outputs.fargate_profile_role_arns
```
per tflint, this is no longer in use
```suggestion
```

dudymas commented on modules/eks/cluster/CHANGELOG.md:
```diff
@@ -1,11 +1,211 @@
+has role attached), you can simply remove the resource from the Terraform state with `[atmos] terraform state rm`,
+because it will be managed by the Karpenter controller instead of Terraform.
+
+### Migration Procedure
```
Assuming this file grows arbitrarily in the future, assume there could be a collision with this title:
```suggestion
### Access Control API Migration Procedure
```

dudymas commented on modules/eks/cluster/CHANGELOG.md:
```diff
@@ -1,11 +1,211 @@
+  `kubectl delete configmap aws-auth --namespace kube-system`. This will not affect the cluster, because it is now being
+  managed by the new access control API, but it will reduce the possibility of confusion in the future.
+
+### End of Migration
```
Assuming this file grows arbitrarily in the future, assume there could be a collision with this title:
```suggestion
### End of Access Control API Migration
```

milldr commented on .pre-commit-config.yaml:
```diff
@@ -44,9 +44,11 @@ repos:
         exclude: |
           (?x)^(
-              README.md |
+              README.md | modules/eks/cluster/README.md |
```
We don't want to ignore any specific file here. Instead, wrap the section you want ignored in the README directly.

See Slack thread respond and this ARB
https://github.com/orgs/cloudposse-corp/discussions/233

dudymas replied:
I've pushed a fix for this now

milldr reviewed (dismissed):
please see comment about prettier

dudymas commented on modules/eks/cluster/README.md:
```diff
@@ -73,40 +153,86 @@ components:
+            # baseload requirements, and in production, we recommend you
+            # have a large enough node group to handle 3/2 (1.5) times your
+            # baseload requirements, to handle the loss of a single AZ.
+            desired_group_size: 3 # number of instances to start with, should be >= number of AZs
+            min_group_size: 3 # must be  >= number of AZs
+            max_group_size: 3
```
This have been a point confusion for multiple end users. If you have 3 AZ's, setting this to 5 or to 3 is effectively rendered the same and has no effect. This is due to [this line in the node group module](https://github.com/cloudposse/terraform-aws-components/blob/main/modules/eks/cluster/modules/node_group_by_region/main.tf#L13-L15)

Really, we should change this variable to be more clear, like `desired_per_az` and `min_per_az`. 

```suggestion
            desired_group_size: 3 # number of instances to start with, should be a multiple of the # of AZs you have
            min_group_size: 3 # must be a multiple of the # of AZs you have
            max_group_size: 3 # must be a multiple of the # of AZs you have
```

Nuru replied:
@dudymas You are mistaken. 

When you use 3 AZs:
- If you set `max_group_size` to 3, you get 3 node groups/ASGs (one per AZ), each with a `max_size` of 1
- If you set `max_group_size` to 5, you get 3 node groups/ASGs (one per AZ), 1 of which has a `max_size` of 1, while the other 2 have `max_size` of 2

dudymas replied:
Desk-checked this again. You're right, and I missed the way index was getting added. As for the behavior in the past, it must have been a different module I'm getting this confused with.

dudymas commented on modules/eks/cluster/variables.tf:
```diff
@@ -93,27 +93,11 @@ variable "cluster_log_retention_period" {
-  nullable    = false
-}
-
-variable "map_additional_worker_roles" {
-  type        = list(string)
-  description = "AWS IAM Role ARNs of worker nodes to add to `aws-auth` ConfigMap"
-  default     = []
-  nullable    = false
-}
-
```
The variable `map_additional_worker_roles` is still used in your main.tf, so this needs to be restored

dudymas reviewed (changes requested):
Please look at the issue with `map_additional_worker_roles`. It was removed from `variables.tf` but is still in use in `main.tf`. I believe @Benbentwo last touched that variable, so he might have more context around its usage/intent.

Nuru commented:
> Please look at the issue with `map_additional_worker_roles`. It was removed from `variables.tf` but is still in use in `main.tf`. I believe @Benbentwo last touched that variable, so he might have more context around its usage/intent.

I had neglected to add the new `variables-deprecated.tf` file, which is where I had moved that variable to. Fixed now.

dudymas reviewed (approved):
LGTM
